### ISSUE TITLE
fetch_member is not optional, despite documentation

### ISSUE THREAD

**Issue by nekokatt** [Rapptz/discord.py#2473]
I am not sure if this is a bug, or a misunderstanding of what `Optional` implies, which I assume this library is following, but this is assuming that `Optional[T]` is used with the same meaning as `typing.Optional[T]` and is thus the same as `typing.Union[T, None]` as per the Python documentation.

Suppose I make the following call in 1b9108f:

```py
guild_obj: discord.Guild = ...

# I want to fetch all members in the guild.
async for member_obj in guild_obj.fetch_members(limit=None):
    ...
```
I will receive the following error on execution:
```
Ignoring exception in on_guild_available
Traceback (most recent call last):
  File "/usr/local/src/discord.py/discord/client.py", line 312, in _run_event
    await coro(*args, **kwargs)
  File "/neko3/neko3/features/autorole.py", line 51, in on_guild_available
    async for member_obj in guild_obj.fetch_members(limit=None):
  File "/usr/local/src/discord.py/discord/iterators.py", line 86, in __anext__
    msg = await self.next()
  File "/usr/local/src/discord.py/discord/iterators.py", line 607, in next
    await self.fill_members()
  File "/usr/local/src/discord.py/discord/iterators.py", line 615, in fill_members
    if self.limit > 0:
TypeError: '>' not supported between instances of 'NoneType' and 'int'
```
However, [the documentation for `1b9108f`](https://github.com/Rapptz/discord.py/blob/1b9108ffd75fb257b699d1af41f0b9c22fc04a97/discord/guild.py#L1210)  states that this is an `Optional`, so I assume that the above example should work correctly.

Side note, I am aware that in most cases, the `members` attribute would provide the same behaviour without the network overhead, and thus using that would be a work around for this.

**Comment by BluePhoenixGame** [comment]
Optional in this case I think means you don't need to pass that arg/kwarg. So you could just have nothing inside the parentheses. But I believe it's a bug if it breaks when you pass in None.
As you can see it's trying to see if you added a limit by seeing if the limit is bigger than 0. I believe it says on the docs that 0 means no limit.

**Comment by nekokatt** [comment]
@BluePhoenixGame If you pass no args, it defaults to fetching a single member (`limit=1`).

If the definition of `Optional` here differs to `typing.Optional`, this should probably be stated in a preface for the documentation so that people do not confuse it with `typing.Optional`, which is generally what people would assume given it appears to be a type hint. Especially since providing the actual `typing.Optional` class in Sphinx in a `:class:` tag would produce the same output, this feels somewhat ambiguous.

But yeah, I can see that. I would have put a PR up to address this but I don't know what the behaviour is meant to be, so I held off doing that. 

Also, at the link I provided, there is unfortunately no mention of 0 being a valid default. 

It would appear that [the _List Guild Members_ endpoint](https://discordapp.com/developers/docs/resources/guild#list-guild-members) does not support optionality for that argument, or no limits. I assume this is the endpoint this call is hitting.

In this case, the documentation should also be amended to specify the limit.

TLDR:
    - Make it clear that `Optional` is not implying the same thing as `typing.Optional` in the "type hint-looking" fields on Sphinx.
    - Specify 1 as the minimum limit and 1000 as the maximum limit as per the link above if it is using the list guild members endpoint.

**Comment by BluePhoenixGame** [comment]
Actually I think it *should* be the same as `typing.Optional`, but a mistake was made here. The rest of the documentation I believe work like `typing.Optional`.
Which should be fixed obviously. I also think that the default should be changed to 1000. If you don't have a limit you basically want to retrieve as many members as possible.

**Comment by nekokatt** [comment]
Yeah, i agree. Seems strange from a REST API perspective that it caps at 1,000 rather than being paginated. I believe the default is 1 though because thats what the API itself defaults to

**Comment by skyblockz** [comment]
It's your problem, optional means you either fill with the required value type, or do not fill it
if you dont fill it, the limit will just be 1
filling None does not mean optional
https://github.com/Rapptz/discord.py/blob/2770137bd6d53dbfbeb325f74f302cf1ffd26f99/discord/guild.py#L1133

**Comment by Rapptz** [comment]
This is a bug and it'll be fixed.
